indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/ParallelIndexSupervisorTask.java

```diff
@@ -901,13 +906,43 @@ public static Map<Interval, Integer> determineNumShardsFromCardinalityReport(
   {
     // aggregate all the sub-reports
     Map<Interval, Union> finalCollectors = mergeCardinalityReports(reports);
+    return computeIntervalToNumShards(maxRowsPerSegment, finalCollectors);
+  }
 
+  @Nonnull
+  @VisibleForTesting
+  static Map<Interval, Integer> computeIntervalToNumShards(
+      int maxRowsPerSegment,
+      Map<Interval, Union> finalCollectors
+  )
+  {
     return CollectionUtils.mapValues(
         finalCollectors,
         union -> {
           final double estimatedCardinality = union.getEstimate();
-          // determine numShards based on maxRowsPerSegment and the cardinality
-          final long estimatedNumShards = Math.round(estimatedCardinality / maxRowsPerSegment);
+          final long estimatedNumShards;
+          if (estimatedCardinality <= 0) {
+            // I don't think we can use the estimate in any way being negative, seven sounds like a nice prime number
+            // it is ok if we end up not filling them all, the ingestion code handles that
+            // Seven on the other hand will at least create some shards rather than potentially a single huge one
+            estimatedNumShards = 7L;
```

Comment:
instead of setting this 7 here, we should move this as a final static variable up top. Something like `DEFAULT_NUM_SHARDS`

Reply:
Just a question, why we went with 7 and say why not 3 ? Any rationale behind it ?

Reply:
Could also make this configurable with a default of 7.

Reply:
My thinking is that three is too small but it is all arbitrary. This is not really a default, it is just for the *remote*, never actually observed, case that estimate is negative. So I would like to leave as is.

Reply:
But in the case that we get into this situation, should the fall back here be configurable, so that if 7 results in a bad estimate / default value, the job can be rerun with a different value and produce potentially better results?

Reply:
If we fall again here we should collect evidence and fix it for good.

Reply:
In order to enforce collect and fix we could also throw an ISE here so the context is repeatable...how does this sound instead of the guesstimate of seven shards? Rather than guesstimating just throw an ISE and halt. This may be too harsh so the warning is better I think but stop there and not try to be more clever. Let's think of this as some sort of fishing expedition for data to see if this was the original problem. There is no evidence and those of us that have tried have not been able to reproduce the scenario.

Reply:
hmm, yeah failing the ingestion completely may be extreme. I think best is to allow user to set a fallback value if it cant be found, maybe a negative value for the config could indicate to fail the ingestion?

Reply:
I realize that complicates things, so your choice to add or not. Would be good to allow some control here, but not a blocker.

Reply:
Put the magic number behind a label.